### PR TITLE
[IOTDB-3104] Add Consensus Module StateMachine Event API

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/IStateMachine.java [hunk 1]
```diff
@@ -67,4 +70,40 @@ interface Registry extends Function<ConsensusGroupId, IStateMachine> {}
...
+  /** An optional API for event notifications. */
+  interface EventApi {
+    /**
+     * Notify the {@link IStateMachine} that a new leader has been elected. Note that the new leader
+     * can possibly be this server.
+     *
+     * @param groupId The id of this consensus group.
+     * @param newLeader The id of the new leader.
+     */
+    default void notifyLeaderChanged(ConsensusGroupId groupId, TEndPoint newLeader) {}
```
Comment (wangchao316) [inline consensus/src/main/java/org/apache/iotdb/consensus/IStateMachine.java:83]: Good

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java [hunk 2]
```diff
@@ -159,4 +168,23 @@ private void loadSnapshot(File latestSnapshotDir) {
...
+
+  public void notifyLeaderChanged(RaftGroupMemberId groupMemberId, RaftPeerId newLeaderId) {
+    applicationStateMachine
+        .event()
+        .notifyLeaderChanged(
+            Utils.fromRaftGroupIdToConsensusGroupId(groupMemberId.getGroupId()),
+            Utils.formRaftPeerIdToTEndPoint(newLeaderId));
+  }
+
+  public void notifyConfigurationChanged(
```
Comment (SzyWilliam) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:180]: The same as above

**Review by SzyWilliam — APPROVED** [review]
LGTM~

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java [hunk 3]
```diff
@@ -159,4 +168,23 @@ private void loadSnapshot(File latestSnapshotDir) {
   public StateMachineStorage getStateMachineStorage() {
     return snapshotStorage;
   }
+
```
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:171]: Fixed

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java [hunk 4]
```diff
@@ -136,11 +141,15 @@ public long takeSnapshot() throws IOException {
     // require the application statemachine to take the latest snapshot
     String metadata = Utils.getMetadataFromTermIndex(lastApplied);
     File snapshotDir = snapshotStorage.getSnapshotDir(metadata);
+    snapshotDir.mkdir();
```
Comment (wangchao316) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:144]: if snapshotDir is exist ,  will make dir?

Comment (OneSizeFitsQuorum) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:144]: Then 'snapshotdir.mkdir ()' will return false and the directory still exists

**Review by wangchao316 — APPROVED** [review]
LGTM

**Pull request merged**